### PR TITLE
fix: revert decoder to Normal state on decoding error

### LINKED ISSUE #3: Panic 'byte index 1 is out of bounds of ``'
Failing test:

```
    #[test]
    fn test_decode_panic() {
        let bytes = b"&#1
 ;";
        escaper::decode_html_buf_sloppy(bytes).ok();
    }
```

```
running 1 test
thread 'mimeparser::tests::test_decode_panic' panicked at 'byte index 1 is out of bounds of ``', /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/escaper-0.1.0/src/decode.rs:168:53
stack backtrace:
   0: rust_begin_unwind
             at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/core/src/panicking.rs:92:14
   2: core::str::slice_error_fail
   3: core::str::traits::<impl core::slice::index::SliceIndex<str> for core::ops::range::RangeFrom<usize>>::index
   4: core::str::traits::<impl core::ops::index::Index<I> for str>::index
   5: <alloc::string::String as core::ops::index::Index<core::ops::range::RangeFrom<usize>>>::index
   6: escaper::decode::decode_html_rw
   7: escaper::decode::decode_html_buf_sloppy
```

### PR DESCRIPTION
**Pull request by link2xt**
Otherwise already cleared `buf` may be used as if decoder is in the
middle of parsing an entity.

Fixes #3

### CODE REVIEW

**Comment by link2xt** [comment]
First commit fixes the bug.

Second commit forbids indexing and slicing in clippy, which is not necessary after the fix, but would convert other similar bugs into errors rather than panics if there are still code paths in the state machine which lead to invalid states.

File: src/decode.rs [hunk 1]
```diff
@@ -165,21 +167,28 @@ pub fn decode_html_rw<R: BufRead, W: Write>(
...
                 try_dec_io!(write_char(writer, ch), good_pos);
                 buf.clear();
             }
             Hex if is_hex_digit(c) => buf.push(c),
             Dec if is_digit(c) => buf.push(c),
             Numeric | Hex | Dec => {
                 if sloppy {
-                    buf.clear()
+                    buf.clear();
+                    state = Normal;
```
Comment (link2xt) [inline src/decode.rs:191]: This is the core of the fix

File: src/decode.rs [hunk 2]
```diff
@@ -165,21 +167,28 @@ pub fn decode_html_rw<R: BufRead, W: Write>(
             Numeric if c == 'x' => state = Hex,
             Dec if c == ';' => {
                 state = Normal;
-                let ch = try_parse!(decode_numeric(&buf[1..], 10), good_pos);
+                let ch = try_parse!(
+                    decode_numeric(&buf.get(1..).unwrap_or_default(), 10),
```
Comment (Jikstra) [inline src/decode.rs:171]: Ohh that's the safe rusty way of indexing.
In the or_default() case, decode_numeric would get what exactly?

File: src/decode.rs [hunk 3]
```diff
@@ -275,3 +284,16 @@ fn decode_numeric(esc: &str, radix: u32) -> Result<char, DecodeErrKind> {
...
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    // Regression test for "out of bounds" panic in sloppy decoder.
+    #[test]
+    fn test_decode_panic() {
+        let bytes = b"&#1 ;foobar";
```
Comment (Jikstra) [inline src/decode.rs:295]: Did the same panic happen with a whitespace? (I guess yes, non digit/hex character after we go in the Numeric | Dec | Hex state)

**Comment by Jikstra** [comment]
> Second commit forbids indexing and slicing in clippy, which is not necessary after the fix, but would convert other similar bugs into errors rather than panics if there are still code paths in the state machine which lead to invalid states.

Nice, i'm suprised that indexing with panics is a thing in rust. Any ideas why? Performance reasons?

Thanks for this PR, learned some things just by trying myself and seeing the proper fix later :)

File: src/decode.rs (hunk 2 above)
Comment (link2xt) [reply src/decode.rs:171]: > In the or_default() case, decode_numeric would get what exactly?

Empty string. Then `decode_numeric` will return an error.

This should never happen anyway. It is not a part of the fix, but a second "defensive" commit.

File: src/decode.rs (hunk 3 above)
Comment (link2xt) [reply src/decode.rs:295]: Yes, just as newline it's invalid and clears `buf`.

**Comment by link2xt** [comment]
> Nice, i'm suprised that indexing with panics is a thing in rust. Any ideas why? Performance reasons?

Early design mistake I think. When this was designed Rust did not have usable errors yet, no `?` operator etc. It was just "safe at all costs", throwing a panic for any exception. When Rust was used in practice (Servo browser engine mostly) it became obvious that crashing apps is not an option, so many alternative implementations, like `try_...` returning a `Result` instead of panic were added to the standard library and a common practice is to forbid many operations like `unwrap`, `expect`, indexing, slicing, `as`, some parts of arithmetics (use `Wrapping`, `.wrapping_add` or `.checked_add` instead) and so on.

**Review by dignifiedquire — APPROVED** [review]

**Pull request merged**